### DISCUSSION TITLE
Efficiency Corrected Power Input Sensor

### DISCUSSION THREAD

**Discussion by JensenNick** [JensenNick/huawei_solar_pees#8] [Forum]
In connection with the latest update of my *"Huawei Solar PEES package"* I've also included a new efficiency corrected power input sensor. The sensor is based upon f(x) functions which have been tweaked to represent the efficiency curves in the Huawei data sheets. This will hopefully result in a more accurate calculation of the yield, a more easy way to tweak it to fit and more smooth handeling of the efficiency.

Is it perfect? No it is an approximation but with very little deviation. You can check out the f(x) formulas and se the approximation done here. 
* [Huawei SUN2000 3/4/5/6/8/10KTL-M1 (triple phase) inverters](https://www.desmos.com/calculator/6fkt4nlr2k)
* [Huawei SUN2000 2/3/3.68/4/4.6/5/6KTL-L1 (single phase) inverters](https://www.desmos.com/calculator/zsdg5o2ykg)

Please refer to the [Wiki Pages](https://github.com/JensenNick/huawei_solar_pees/wiki/1.-Power-Sensors#efficiency-corrected-inverter-power-sensors) for a more detailed description and the code itself.

![efficiency_graphs](https://github.com/JensenNick/huawei_solar_pees/assets/149193835/a979b2a5-277c-4956-86eb-0759b9607d32)

![efficiency_corrected_sensor](https://github.com/JensenNick/huawei_solar_pees/assets/149193835/9909068b-b75f-41c4-b462-bf6e087a6c6d)

**Comment by Roving-Ronin** [comment]
Here'e the efficiency calculation I'm using with 2 x M1 inverters. Someone said they had issues when then connected with a battery (WLCRS not giving the efficiency sensor data), but should be able to test with another users 2 x 5kW L1 and 15kW LUNA soon.

```
  # Shows 'Daily Solar Yield' being the yield of your solar panels, minus any efficiency losses by the inverter
  # See: https://github.com/wlcrs/huawei_solar/wiki/Daily-Solar-Yield
  - sensor:
    - name: "Inverters Input Power with Efficiency Loss"
      unique_id: inverters_input_power_with_efficiency_loss
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      state: >
        {% set inverter1_input_with_efficiency = (states('sensor.inverter1_efficiency') | float(0) /100) * states('sensor.inverter1_input_power') | float(0) %}
        {% set inverter2_input_with_efficiency = (states('sensor.inverter2_efficiency') | float(0) /100) * states('sensor.inverter2_input_power') | float(0) %}
        {{ (inverter1_input_with_efficiency + inverter2_input_with_efficiency) | round(3) }}
      availability: >-
        {{ (states('sensor.inverter1_efficiency')|is_number)
            and (states('sensor.inverter2_efficiency')|is_number) 
              and (states('sensor.inverter1_input_power')|is_number) 
                and (states('sensor.inverter2_input_power')|is_number) }}
 
```

**Reply by JensenNick** [reply]
@Roving-Ronin Thanks! Yes the issue is that the state of `sensor.inverter_efficiency` is returning 100% and it never changes, when/if a Huawei LUNA battery is connected to the inverter.
![2024-01-26_11-43-33](https://github.com/JensenNick/huawei_solar_pees/assets/149193835/ee0cb102-537a-4700-8412-cd0437f02d62)

**Reply by Roving-Ronin** [reply]
@JensenNick just got access to another users 2 x L1 5kW inverters, with the primary inverter having a 15kWh LUNA ESS attached.
Checking the primary inverter and 99% of the time its efficiency is showing 100% (so it would appear this is due to having a battery attached), whilst I do note that the secondary inverter (no battery) does have its efficiency sensor changing throughout the day as the sun comes up / goes down. 9:30am where they are located (Perth) and Primary inverter showing 100%, secondary inverter showing 97.99% efficiency.
